# Exercícios: Aspectos Nativos de Python 🐍

## Manipulação de variáveis e funções básicas

Na aula de hoje, você aprendeu sobre tipos de variáveis e como utilizar funções básicas do Python. Responda aos exercícios conforme as instruções.

😇 **Exercício 1**: Olá mundo!
1. Crie duas variáveis chamadas `nome` e `sobrenome` e atribua o seu nome e o seu sobrenome a elas, respectivamente.

In [1]:
nome = 'kalina'
sobrenome = 'porto'

2. Utilize funções de manipulação de *strings* para capitalizar a primeira letra do seu nome e transformar em maiúsculas todas as letras do seu sobrenome.

In [8]:
nome = nome.capitalize()
sobrenome = sobrenome.upper()

3. Imprima, na mesma linha, a seguinte frase, utilizando o conteúdo das variáveis nos locais indicados:
<br><br> Olá Mundo! Meu nome é `sobrenome`, `nome` `sobrenome`

In [9]:
print(f'Olá mundo! Meu nome é {sobrenome}, {nome} {sobrenome}')

Olá mundo! Meu nome é PORTO, Kalina PORTO


4. Agora... Troque os valores de `nome` e `sobrenome` sem usar variáveis adicionais! Imprima os valores de `nome` e `sobrenome` para se certificar de que elas foram trocadas.

In [10]:
nome, sobrenome = sobrenome, nome
print(f'nome = {nome}, sobrenome = {sobrenome}')

nome = PORTO, sobrenome = Kalina


## Estruturas de repetição

😇 **Exercício 2**: Irmãozinho chato!

Crie um pequeno código que simule aquele irmãozinho irritante! O código deverá ler da entrada uma frase inteira do tipo *Sai do meu quarto!* e imprimir a mesma frase com as palavras `nhem nhem! :-P` no final. A brincadeira deverá continuar até que o usuário digite 'mamãe!', ou seja, quando o usuário digitar 'mamae!', o código deverá finalizar sem a repetição da última frase. Exemplo de execução do código:

<br>
<br> Entrada: Sai do meu quarto!
<br> Sai do meu quarto! nhem nhem! :-P
<br> Entrada: Vou chamar a mamãe!
<br> Vou chamar a mamãe! nhem nhem! :-P
<br> Entrada: mamãe!


In [12]:
entrada = input('Diga algo: ')
while(entrada != 'mamãe!'):
    print(f'{entrada} nhem nhem! :-P')
    entrada = input('Diga algo: ')

Diga algo: sai daqui!
sai daqui! nhem nhem! :-P
Diga algo: mamãe!


😯 **Exercício 3**: Analizador de sequência de números

Escreva um programa que analise uma sequência de números inteiros para identificar certas características. O programa deverá ser capaz de detectar se a sequência é *crescente*, *decrescente*, ou *sem padrão*. Ele também deve calcular a média dos números lidos e determinar quantos números estão acima e abaixo da média.

Passo 1: Pergunte ao usuário o número **n** de elementos que ele deseja na sequência.

In [32]:
n = int(input('Digite o número de elementos desejados: '))

Digite o número de elementos desejados: 5


Passo 2: Usando uma estrutura de repetição, solicite ao usuário para prover **n** números de entrada, um a um. Armazene esses números em uma lista.
<br>
*Dica*: Para adicionar elementos a uma lista, utilize a função **append()**.

In [33]:
seq = []
for i in range(n):
    temp = int(input(f'Digite o número {i+1}: '))
    seq.append(temp)

Digite o número 1: 4
Digite o número 2: 5
Digite o número 3: 6
Digite o número 4: 7
Digite o número 5: 8


Passo 3: Analise a sequência de números para determinar se ela é *crescente*, *decrescente* ou *sem padrão*. 
<br>
*Dica*: Use estruturas condicionais para determinar se a sequência é crescente ou decrescente.

In [34]:
#Usando a função sorted do Python.
if seq == sorted(seq):
    print('A sequência é crescente.')
elif seq == sorted(seq,reverse=True):
    print('A sequência é decrescente.')
else:
    print('A sequência não tem padrão.')

A sequência é crescente.


Passo 4: Calcule a média da sequência de números e imprima o resultado.

In [35]:
media = sum(seq)/n
print(f'A média da sequência é {media:.2f}')

A média da sequência é 6.00


Passo 5: Determine as quantidades de números acima e abaixo da média e imprima os resultados.

In [36]:
abaixo_media = 0
acima_media = 0
for i in seq:
    if i < media:
        abaixo_media +=1
    elif i > media:
        acima_media +=1
print(f'Existem {abaixo_media} números abaixo da média.')
print(f'Existem {acima_media} números acima da média.')

Existem 2 números abaixo da média.
Existem 2 números acima da média.


## 🤯 Desafio! Sequência de Collatz

A sequência de Collatz é uma sequência de números definida da seguinte forma: Comece com qualquer número inteiro positivo **n**. Se **n** é par, divida-o por 2; se **n** é ímpar, multiplique-o por 3 e adicione 1. Repita o processo com o novo valor de **n**, terminando a sequência quando **n = 1**.
<br>

O objetivo desse desafio é explorar as propriedades e padrões dentro de uma sequência de Collatz para números inteiros até um determinado limite **m**.
<br></br>
**Problema**: Dada uma sequência de números positivos de 1 a **m**:
1. Gere a sequência de Collatz para cada um dos números da sequência.
2. Determine qual o número da sequência produz a sequência de Collatz mais longa.
3. Calcule o tamanho médio das sequências de Collatz para números da sequência dada.
4. Dado um número específico **k**, verifique quantas sequências passam por este número antes de atingir 1.


Passo 1: Solicite ao usuário o limite **m** da sequência.

In [38]:
# Get input values
m = int(input("Forneça o limite da sequência (m): "))

Forneça o limite da sequência (m): 1000


Passo 2: Gere a sequência de Collatz para cada número de 1 a **m**. Armazene o tamanho de cada sequência.

In [45]:
seq_mais_longa = 0
num_seq_mais_longa = 0
tam_total = 0

for i in range(1, m+1):
    n = i
    tam = 0
    
    while n != 1:
        if n % 2 == 0:
            n = n // 2
        else:
            n = 3 * n + 1
        tam += 1

    tam_total += tam

    if tam > seq_mais_longa:
        seq_mais_longa = tam
        num_seq_mais_longa = i

Passo 3: Determine qual número produz a sequência mais longa e o tamanho médio das sequências.

In [46]:
tam_medio = tam_total / m

Passo 4: Solicite ao usuário um número **k** qualquer e verifique se alguma sequência passa por esse número.

In [47]:
k = int(input("Forneça um número (k) para verificar se ele está em alguma sequência: "))

Forneça um número (k) para verificar se ele está em alguma sequência: 524


In [48]:
k_count = 0

for i in range(1, m+1):
    n = i
    
    while n != 1:
        if n == k:
            k_count += 1
            break
        if n % 2 == 0:
            n = n // 2
        else:
            n = 3 * n + 1

Passo 5: Imprima os resultados.
<br></br>
Exemplo: 
<br></br>(Entrada) 
<br></br>Forneça o limite da sequência (m): **1000**
<br></br>Forneça um número (k) para verificar se ele está em alguma sequência: **524**
<br></br>(Saída) 
<br></br>O número 871 produz a sequência de Collatz mais longa, de tamanho 179.
<br></br>A média de tamanho das sequências de Collatz produzidas para números até 1000 é 62,8. 
<br></br>O número 524 aparece em 14 sequências.

In [49]:
# Print results
print(f"O número {num_seq_mais_longa} produz a sequência de Collatz mais longa, com tamanho {seq_mais_longa}.")
print(f"A média de tamanho das sequências de Collatz produzidas para números até {m} is {tam_medio:.3f}.")
print(f"O número {k} aparece em {k_count} sequências.")


O número 871 produz a sequência de Collatz mais longa, com tamanho 178.
A média de tamanho das sequências de Collatz produzidas para números até 1000 is 59.542.
O número 524 aparece em 14 sequências.
